GPU/Common/GPUStateUtils.cpp

```diff
@@ -386,3 +408,627 @@ LogicOpReplaceType ReplaceLogicOpType() {
+		// No viewport transform here. Let's experiment with using region.
+		out.viewportX = renderX + displayOffsetX;
+		out.viewportY = renderY + displayOffsetY;
+		out.viewportW = curRTWidth * renderWidthFactor;
+		out.viewportH = curRTHeight * renderHeightFactor;
+		out.depthRangeMin = 0.0f;
+		out.depthRangeMax = 1.0f;
+	} else {
+		// These we can turn into a glViewport call, offset by offsetX and offsetY. Math after.
+		float vpXScale = gstate.getViewportXScale();
+		float vpXCenter = gstate.getViewportXCenter();
+		float vpYScale = gstate.getViewportYScale();
+		float vpYCenter = gstate.getViewportYCenter();
+
+		// The viewport transform appears to go like this:
+		// Xscreen = -offsetX + vpXCenter + vpXScale * Xview
+		// Yscreen = -offsetY + vpYCenter + vpYScale * Yview
+		// Zscreen = vpZCenter + vpZScale * Zview
+
+		// The viewport is normally centered at 2048,2048 but can also be centered at other locations.
+		// Offset is subtracted from the viewport center and is also set to values in those ranges, and is set so that the viewport will cover
+		// the desired screen area ([0-480)x[0-272)), so 1808,1912.
+
+		// This means that to get the analogue glViewport we must:
+		float vpX0 = vpXCenter - offsetX - fabsf(vpXScale);
+		float vpY0 = vpYCenter - offsetY - fabsf(vpYScale);   // Need to account for sign of Y
+		gstate_c.vpWidth = vpXScale * 2.0f;
+		gstate_c.vpHeight = vpYScale * 2.0f;
+
+		float vpWidth = fabsf(gstate_c.vpWidth);
+		float vpHeight = fabsf(gstate_c.vpHeight);
+
+		// This multiplication should probably be done after viewport clipping. Would let us very slightly simplify the clipping logic?
+		vpX0 *= renderWidthFactor;
+		vpY0 *= renderHeightFactor;
+		vpWidth *= renderWidthFactor;
+		vpHeight *= renderHeightFactor;
+
+		// We used to apply the viewport here via glstate, but there are limits which vary by driver.
+		// This may mean some games won't work, or at least won't work at higher render resolutions.
+		// So we apply it in the shader instead.
+		float left = renderX + vpX0;
+		float top = renderY + vpY0;
+		float right = left + vpWidth;
+		float bottom = top + vpHeight;
+
+		float wScale = 1.0f;
+		float xOffset = 0.0f;
+		float hScale = 1.0f;
+		float yOffset = 0.0f;
+
+		// If we're within the bounds, we want clipping the viewport way.  So leave it be.
+		if (left < 0.0f || right > renderWidth) {
+			float overageLeft = std::max(-left, 0.0f);
+			float overageRight = std::max(right - renderWidth, 0.0f);
+			// Our center drifted by the difference in overages.
+			float drift = overageRight - overageLeft;
+
+			left += overageLeft;
+			right -= overageRight;
+
+			wScale = vpWidth / (right - left);
+			xOffset = drift / (right - left);
+		}
+
+		if (top < 0.0f || bottom > renderHeight) {
+			float overageTop = std::max(-top, 0.0f);
+			float overageBottom = std::max(bottom - renderHeight, 0.0f);
+			// Our center drifted by the difference in overages.
+			float drift = overageBottom - overageTop;
```

Comment:
This was reversed in D3D9; I'm pretty sure that's wrong, but do we need to negate vpYOffset now inside ShaderManagerDX9?  Without doing that, Star Ocean is broken in Direct3D 9 with these changes.

To be fair, `-gstate_c.vpYOffset` seems wrong and so I don't understand why it works...

-[Unknown]